src/opa.rs

```diff
@@ -0,0 +1,245 @@
+//! This module offers common access to the [`OpaConfig`] which can be used in operators
+//! to specify a name for a [`k8s_openapi::api::core::v1::ConfigMap`] and a package name
+//! for OPA rules.
+//!
+//! Additionally several methods are provided to build an URL to query the OPA data API.
+//!
+use kube::ResourceExt;
+use schemars::{self, JsonSchema};
+use serde::{Deserialize, Serialize};
+
+const OPA_API: &str = "v1/data/";
+
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct OpaConfig {
+    pub config_map_name: String,
+    pub package: Option<String>,
+}
+
+impl OpaConfig {
+    /// Returns the full qualified OPA data API url up to the package. If [`OpaConfig`] has
+    /// no `package` set, will default to the cluster `resource` name.
+    ///
+    /// # Example
+    ///
+    /// http://localhost:8080/v1/data/<package>
+    ///
+    /// # Arguments
+    /// * `resource`     - The cluster resource
+    /// * `opa_base_url` - The base url to OPA e.g. http://localhost:8081
+    pub fn full_package_url<T>(&self, resource: &T, opa_base_url: &str) -> String
+    where
+        T: ResourceExt,
+    {
+        if opa_base_url.ends_with('/') {
+            format!("{}{}", opa_base_url, self.package_url(resource))
+        } else {
+            format!("{}/{}", opa_base_url, self.package_url(resource))
+        }
+    }
+
+    /// Returns the full qualified OPA data API url up to the rule. If [`OpaConfig`] has
+    /// no `package` set, will default to the cluster `resource` name.
+    ///
+    /// # Example
+    ///
+    /// http://localhost:8080/v1/data/<package>/<rule>
+    ///
+    /// # Arguments
+    /// * `resource`     - The cluster resource.
+    /// * `opa_base_url` - The base url to OPA e.g. http://localhost:8081.
+    /// * `rule`         - The rule name. Defaults to `allow`.
+    pub fn full_rule_url<T>(&self, resource: &T, opa_base_url: &str, rule: Option<&str>) -> String
+    where
+        T: ResourceExt,
+    {
+        if opa_base_url.ends_with('/') {
+            format!("{}{}", opa_base_url, self.rule_url(resource, rule))
+        } else {
+            format!("{}/{}", opa_base_url, self.rule_url(resource, rule))
+        }
+    }
+
+    /// Returns the OPA data API url up to the package. If [`OpaConfig`] has
+    /// no `package` set, will default to the cluster `resource` name.
+    ///
+    /// This may be used if the OPA base url is contained in an ENV variable.
+    ///
+    /// # Example
+    ///
+    /// v1/data/<package>
+    ///
+    /// # Arguments
+    /// * `resource`     - The cluster resource.
+    pub fn package_url<T>(&self, resource: &T) -> String
```

Comment:
Personally I'm not a big fan of this: having a (potentially) complex resource as a dependency (parameter) just to read one field from it is usually an anti pattern. Maybe we can introduce a `ClusterName` type or something like that.

Reply:
Its just something of type `ResourceExt` for the name() and namespace() methods. It defines the behavior as in the doc that we default to the resource name. That would not be true if we have a parameter `cluster_name`.